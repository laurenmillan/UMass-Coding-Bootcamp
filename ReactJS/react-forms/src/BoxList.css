.BoxList {
	font-family: 'Franklin Gothic Medium', 'Arial Narrow', Arial, sans-serif;
	justify-content: center;
	align-items: center;
	padding: 50px;
	margin: 100px;
	display: flex;
	text-align: center;
}
