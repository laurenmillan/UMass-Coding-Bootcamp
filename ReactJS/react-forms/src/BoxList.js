import React, { useState } from 'react';
import './BoxList.css';
import Box from './Box';
import NewBoxForm from './NewBoxForm';

/** Color Box Maker.
 *
 * State: boxes, setBoxes
 * Props: addBox, width, height, color
 *
 * - contains all of the boxes
 * - should render all of the Box components along
 *      with the NewBoxForm component
 *
 **/

// initial value of boxes is an empty array, so it can be updated
const BoxList = () => {
	const [ boxes, setBoxes ] = useState([]);

	// this function takes formData as an argument, it creates a new array to copy existing boxes array
	const addBox = (formData) => {
		// setBoxes is called with the new array to update State
		setBoxes((boxes) => [ ...boxes, formData ]);
	};

	// removeBox takes index as an argument - the index of the box to be removed from the boxes array
	const removeBox = (index) => {
		// newBoxes creates a new array by using filter by index
		const newBoxes = boxes.filter((box, i) => i !== index);
		// setBoxes is called with the newBoxes array to update State
		setBoxes(newBoxes);
	};

	// form to create new boxes using the NewBoxForm component and addBox as its prop
	// map iterates over the boxes array and renders a Box component for each element in the array
	//  and is giving width, height, color as props, with a key using the array index
	return (
		<div className="BoxList">
			<NewBoxForm addBox={addBox} />
			{boxes.map((box, i) => (
				<Box key={i} width={box.width} height={box.height} color={box.color} removeBox={() => removeBox(i)} />
			))}
		</div>
	);
};

export default BoxList;
