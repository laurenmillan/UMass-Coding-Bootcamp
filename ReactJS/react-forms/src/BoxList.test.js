import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import BoxList from './BoxList';

// SMOKE TEST
it(' renders without crashing', () => {
	render(<BoxList />);
});

// SNAPSHOT TEST
it('it matches snapshot', () => {
	const { asFragment } = render(<BoxList />);
	expect(asFragment()).toMatchSnapshot();
});
