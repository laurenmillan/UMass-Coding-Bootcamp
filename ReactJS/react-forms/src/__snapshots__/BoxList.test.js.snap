// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="width"
      >
        Width:
      </label>
      <input
        id="width"
        name="width"
        type="text"
        value=""
      />
      <br />
      <label
        for="height"
      >
        Height:
      </label>
      <input
        id="height"
        name="height"
        type="text"
        value=""
      />
      <br />
      <label
        for="color"
      >
        Color:
      </label>
      <input
        id="color"
        name="color"
        type="text"
        value=""
      />
      <br />
      <button
        type="submit"
      >
        Add Box
      </button>
    </form>
  </div>
</DocumentFragment>
`;
